presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java

```diff
@@ -309,8 +309,7 @@ public void write(Page page)
         }
 
         // avoid chunk with huge logical size
-        int averageLogicalSizePerRow = estimateAverageLogicalSizePerRow(page);
-        int maxChunkRowCount = max(1, chunkMaxLogicalBytes / max(1, averageLogicalSizePerRow));
+        int maxChunkRowCount = toIntExact(max(1, chunkMaxLogicalBytes / max(1, page.getApproximateLogicalSizeInBytes() / page.getPositionCount())));
```

Comment:
There are two changes here. 
-  Call approximate logical size
-  Calculate whole page

Have you considered to add a new function `page.getApproximateRegionLogicalSizeInBytes(position = 0, length = 100)` ? I am not sure how faster it would be, however it guarantees there will be no regression due to calculating the whole page.

Reply:
@viczhang861 Thanks for reviewing the PR so quickly! `getApproximateRegionLogicalSizeInBytes` is the same fast calculating a region as calculating a whole page. It is over 20x faster than `getLogicalSizeInBytes()` and do not allocate memory for `DictionaryBlock` as shown in https://github.com/prestodb/presto/pull/15272

Let's see the difference for the two functions in DictionaryBlock
```
@Override
    public long getRegionLogicalSizeInBytes(int positionOffset, int length)
    {
        if (positionOffset == 0 && length == getPositionCount() && logicalSizeInBytes >= 0) {
            return logicalSizeInBytes;
        }

        long sizeInBytes = 0;
        long[] seenSizes = new long[dictionary.getPositionCount()];
        Arrays.fill(seenSizes, -1L);
        for (int i = positionOffset; i < positionOffset + length; i++) {
            int position = getId(i);
            if (seenSizes[position] < 0) {
                seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
            }
            sizeInBytes += seenSizes[position];
        }

        if (positionOffset == 0 && length == getPositionCount()) {
            logicalSizeInBytes = sizeInBytes;
        }

        return sizeInBytes;
    }

    @Override
    public long getApproximateRegionLogicalSizeInBytes(int position, int length)
    {
        int dictionaryPositionCount = dictionary.getPositionCount();
        return dictionaryPositionCount == 0 ? 0 : dictionary.getApproximateRegionLogicalSizeInBytes(0, dictionaryPositionCount) * length / dictionaryPositionCount;
    }
```

The bad thing in `getRegionLogicalSizeInBytes` was` long[] seenSizes = new long[dictionary.getPositionCount()];`.  `getApproximateRegionLogicalSizeInBytes` avoids this problem and just has scalar arithmetic operations. 

With this in mind, calculating the size for the whole page is more accurate and advantageous.

Reply:
@yingsu00 Thanks for the explanation !